meson_cpu/x86/meson.build

```
@@ -1,228 +1,102 @@
+  X86_V2.update(args: [{'val': '/arch:SSE4.2', 'match': '/arch:.*'}] + HWY_SSE4_FLAGS)
+  X86_V3.update(args: {'val': '/arch:CORE-AVX2', 'match': '/arch:.*'})
+  X86_V4.update(args: {'val': '/Qx:SKYLAKE-AVX512', 'match': '/[arch|Qx]:.*'})
   AVX512_ICL.update(args: {'val': '/Qx:ICELAKE-CLIENT', 'match': '/[arch|Qx]:.*'})
 endif
 
 if compiler_id == 'intel'
-  clear_m = '^(-mcpu=|-march=)'
   clear_any = '^(-mcpu=|-march=|-x[A-Z0-9\-])'
-  FMA3.update(args: {'val': '-xCORE-AVX2', 'match': clear_m})
-  AVX2.update(args: {'val': '-xCORE-AVX2', 'match': clear_m})
-  AVX512F.update(args: {'val': '-xCOMMON-AVX512', 'match': clear_m})
-  AVX512CD.update(args: {'val': '-xCOMMON-AVX512', 'match': clear_m})
-  AVX512_KNL.update(args: {'val': '-xKNL', 'match': clear_any})
-  AVX512_KNM.update(args: {'val': '-xKNM', 'match': clear_any})
-  AVX512_SKX.update(args: {'val': '-xSKYLAKE-AVX512', 'match': clear_any})
-  AVX512_CLX.update(args: {'val': '-xCASCADELAKE', 'match': clear_any})
-  AVX512_CNL.update(args: {'val': '-xCANNONLAKE', 'match': clear_any})
+  X86_V2.update(args: [{'val': '-xSSE4.2', 'match': clear_any}] + HWY_SSE4_FLAGS)
+  X86_V3.update(args: {'val': '-xCORE-AVX2', 'match': clear_any})
+  X86_V4.update(args: {'val': '-xSKYLAKE-AVX512', 'match': clear_any})
   AVX512_ICL.update(args: {'val': '-xICELAKE-CLIENT', 'match': clear_any})
 endif
 
 if compiler_id == 'msvc'
-  # MSVC compiler doesn't support the following features
-  foreach fet : [AVX512_KNL, AVX512_KNM]
-    fet.update(disable: compiler_id + ' compiler does not support it')
-  endforeach
-  # The following features don't own private FLAGS, however the compiler still
-  # provides ISA capability for them.
-  foreach fet : [
-    SSE3, SSSE3, SSE41, POPCNT, SSE42, AVX, F16C, XOP, FMA4,
-    AVX512F, AVX512CD, AVX512_CLX, AVX512_CNL,
-    AVX512_ICL
-  ]
-    fet.update(args: '')
-  endforeach
-  # MSVC compiler doesn't support the following features independently
-  FMA3.update(implies: [F16C, AVX2])
-  AVX2.update(implies: [F16C, FMA3])
-  AVX512F.update(implies: [AVX2, AVX512CD, AVX512_SKX])
-  AVX512CD.update(implies: [AVX512F, AVX512_SKX])
+  MSVC_SSE4 = cpu_family == 'x86' ? ['/arch:SSE2'] : []
+  MSVC_SSE4 = cc.version().version_compare('>=19.40') ? ['/arch:SSE4.2'] : MSVC_SSE4 
+  X86_V2.update(args: MSVC_SSE4 + HWY_SSE4_FLAGS)
   clear_arch = '/arch:.*'
-  # only available on 32-bit. Its enabled by default on 64-bit mode
-  foreach fet : [SSE, SSE2]
-    if cpu_family == 'x86'
-      fet.update(args: {'val': '/arch:' + fet.get('name'), 'match': clear_arch})
-    else
-      fet.update(args: '')
-    endif
-  endforeach
-  FMA3.update(args: {'val': '/arch:AVX2', 'match': clear_arch})
-  AVX2.update(args: {'val': '/arch:AVX2', 'match': clear_arch})
-  AVX512_SKX.update(args: {'val': '/arch:AVX512', 'match': clear_arch})
+  X86_V3.update(args: {'val': '/arch:AVX2', 'match': clear_arch})
+  X86_V4.update(args: {'val': '/arch:AVX512', 'match': clear_arch})
```

By default, Highway considers AVX512 a broken platform on MSVC. We've managed to support AVX512 on MSVC through universal intrinsics. We have three options here:

1. Keep it as-is: This leads to building AVX2 kernels with `/arch:AVX512`, which gives the compiler the opportunity to optimize (hopefully). Worst case: we'll have duplicate kernels.

2. Disable AVX512 support entirely:
```suggestion
  X86_V4.update(disable: 'not supported by Highway')
```
3.Change the default behavior and force-enable AVX512 on MSVC and lets see how deep the rabbit-hole goes:
```suggestion
  X86_V4.update(args: [{'val': '/arch:AVX512', 'match': clear_arch}, '-DHWY_BROKEN_MSVC=0'])
```

---

I think I'd prefer (2) or (3). If it's really very broken, disabling it seems fine (and low-effort). If it's feasible and there is energy for this, fixing it up in Highway over time seems nice. 

(1) seems worst, if it doesn't do much, I'd much rather take the binary size gains from disabling it completely.

---

We still haven't converted many kernels yet from universal intrinsics to Highway, so adding extra burden isn't a wise decision. Let's go for option (2) for now and investigate later which MSVC versions are compatible with Highway.